src/ring_theory/is_adjoin_root.lean

```diff
@@ -0,0 +1,578 @@
+  obtain ⟨y, hy⟩ := hzw,
+  rw [← sub_eq_zero, ← eval₂_sub, hy, eval₂_mul, hx, zero_mul]
+end
+
+variables (i x) -- To match `adjoin_root.lift`
+
+/-- Lift a ring homomorphism `R →+* T` to `S →+* T` by specifying a root `x` of `f` in `T`,
+where `S` is given by adjoining a root of `f` to `R`. -/
+def lift (h : is_adjoin_root S f) : S →+* T :=
+{ to_fun := λ z, (h.repr z).eval₂ i x,
+  map_zero' := by rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ (map_zero _), eval₂_zero],
+  map_add' := λ z w, begin
+    rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ (h.repr z + h.repr w), eval₂_add],
+    { rw [map_add, map_repr, map_repr] }
+  end,
+  map_one' := by rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ (map_one _), eval₂_one],
+  map_mul' := λ z w, begin
+    rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ (h.repr z * h.repr w), eval₂_mul],
+    { rw [map_mul, map_repr, map_repr] }
+  end }
+
+variables {i x}
+
+@[simp] lemma lift_map (h : is_adjoin_root S f) (z : R[X]) :
+  h.lift i x hx (h.map z) = z.eval₂ i x :=
+by rw [lift, ring_hom.coe_mk, h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ rfl]
+
+@[simp] lemma lift_root (h : is_adjoin_root S f) :
+  h.lift i x hx h.root = x :=
+by rw [← h.map_X, lift_map, eval₂_X]
+
+@[simp] lemma lift_algebra_map (h : is_adjoin_root S f) (a : R) :
+  h.lift i x hx (algebra_map R S a) = i a :=
+by rw [h.algebra_map_apply, lift_map, eval₂_C]
+
+variables [algebra R T] (hx' : aeval x f = 0)
+
+omit hx
+
+variables (x) -- To match `adjoin_root.lift_hom`
+
+/-- Lift the algebra map `R → T` to `S →ₐ[R] T` by specifying a root `x` of `f` in `T`,
+where `S` is given by adjoining a root of `f` to `R`. -/
+def lift_hom (h : is_adjoin_root S f) : S →ₐ[R] T :=
+{ commutes' := λ a, h.lift_algebra_map hx' a,
+  .. h.lift (algebra_map R T) x hx' }
+
+variables {x}
+
+@[simp] lemma coe_lift_hom (h : is_adjoin_root S f) :
+  (h.lift_hom x hx' : S →+* T) = h.lift (algebra_map R T) x hx' := rfl
+
+lemma lift_algebra_map_apply (h : is_adjoin_root S f) (z : S) :
+  h.lift (algebra_map R T) x hx' z = h.lift_hom x hx' z := rfl
+
+@[simp] lemma lift_hom_map (h : is_adjoin_root S f) (z : R[X]) :
+  h.lift_hom x hx' (h.map z) = aeval x z :=
+by rw [← lift_algebra_map_apply, lift_map, aeval_def]
+
+@[simp] lemma lift_hom_root (h : is_adjoin_root S f) :
+  h.lift_hom x hx' h.root = x :=
+by rw [← lift_algebra_map_apply, lift_root]
+
+end lift
+
+end is_adjoin_root
+
+namespace adjoin_root
+
+variables (f)
+
+/-- `adjoin_root f` is indeed given by adjoining a root of `f`. -/
+protected def is_adjoin_root : is_adjoin_root (adjoin_root f) f :=
+{ map := adjoin_root.mk f,
+  map_surjective := ideal.quotient.mk_surjective,
+  ker_map := begin
+    ext,
+    rw [ring_hom.mem_ker, ← @adjoin_root.mk_self _ _ f, adjoin_root.mk_eq_mk,
+        ideal.mem_span_singleton, ← dvd_add_left (dvd_refl f), sub_add_cancel]
+  end,
+  algebra_map_eq := adjoin_root.algebra_map_eq f }
+
+/-- `adjoin_root f` is indeed given by adjoining a root of `f`. -/
```

Comment:
```suggestion
/-- `adjoin_root f` is indeed given by adjoining a root of `f`. If `f` is monic use this rather than `adjoin_root.is_adjoin_root`. -/
```
Only if you agree.

Reply:
I'd write it as:
```suggestion
/-- `adjoin_root f` is indeed given by adjoining a root of `f`. If `f` is monic this is more
powerful than `adjoin_root.is_adjoin_root`. -/
```